--- ./CMakeLists.txt	2024-12-05 20:33:49.000000000 +0100
+++ ./CMakeLists.txt	2025-01-30 22:37:29.522620203 +0100
@@ -39,6 +39,7 @@
 SET ( DEFAULT_X11                         OFF )
 SET ( DEFAULT_PIPEWIRE                    OFF )
 SET ( DEFAULT_PIPEWIRE_EGL                OFF )
+SET ( DEFAULT_AMLOGIC                     OFF )
 SET ( DEFAULT_FRAMEBUFFER                 OFF )
 SET ( DEFAULT_SOUNDCAPWINDOWS             OFF )
 SET ( DEFAULT_SOUNDCAPMACOS               OFF )
@@ -124,6 +125,7 @@
 	SET ( DEFAULT_SOUNDCAPLINUX ON )
 	SET ( DEFAULT_V4L2          ON )
 	SET ( DEFAULT_X11           ON )
+	SET ( DEFAULT_AMLOGIC       ON )
 	SET ( DEFAULT_FRAMEBUFFER   ON )
 	SET ( DEFAULT_PIPEWIRE      ON )
 	SET ( DEFAULT_SPIDEV        ON )
@@ -445,6 +447,13 @@
 option(ENABLE_DX "Enable Windows DirectX 11 system grabber" ${DEFAULT_DX})
 colorMe("ENABLE_DX = " ${ENABLE_DX})
 
+option(ENABLE_AMLOGIC "Enable the AMLOGIC video grabber" ${DEFAULT_AMLOGIC})
+colorMe("ENABLE_AMLOGIC = " ${ENABLE_AMLOGIC})
+
+if(ENABLE_AMLOGIC)
+	set(ENABLE_FRAMEBUFFER ON)
+endif()
+
 option(ENABLE_FRAMEBUFFER "Enable the framebuffer Linux system grabber" ${DEFAULT_FRAMEBUFFER})
 colorMe("ENABLE_FRAMEBUFFER = " ${ENABLE_FRAMEBUFFER})
 
